smart_open/ftp.py

```diff
@@ -0,0 +1,122 @@
+
+URI_EXAMPLES = (
+    "ftp://username@host/path/file",
+    "ftp://username:password@host/path/file",
+    "ftp://username:password@host:port/path/file",
+)
+
+
+def _unquote(text):
+    return text and urllib.parse.unquote(text)
+
+
+def parse_uri(uri_as_string):
+    split_uri = urllib.parse.urlsplit(uri_as_string)
+    assert split_uri.scheme in SCHEME
+    return dict(
+        scheme=split_uri.scheme,
+        uri_path=_unquote(split_uri.path),
+        user=_unquote(split_uri.username),
+        host=split_uri.hostname,
+        port=int(split_uri.port or DEFAULT_PORT),
+        password=_unquote(split_uri.password),
+    )
+
+
+def open_uri(uri, mode, transport_params):
+    smart_open.utils.check_kwargs(open, transport_params)
+    parsed_uri = parse_uri(uri)
+    uri_path = parsed_uri.pop("uri_path")
+    parsed_uri.pop("scheme")
+    return open(uri_path, mode, transport_params=transport_params, **parsed_uri)
+
+
+def convert_transport_params_to_args(transport_params):
+    supported_keywords = [
+        "timeout",
+        "source_address",
+        "encoding",
+    ]
+    unsupported_keywords = [k for k in transport_params if k not in supported_keywords]
+    kwargs = {k: v for (k, v) in transport_params.items() if k in supported_keywords}
+
+    if unsupported_keywords:
+        logger.warning(
+            "ignoring unsupported ftp keyword arguments: %r", unsupported_keywords
+        )
+
+    return kwargs
+
+
+def _connect(hostname, username, port, password, transport_params):
+    kwargs = convert_transport_params_to_args(transport_params)
+    ftp = FTP(**kwargs)
+    try:
+        ftp.connect(hostname, port)
+    except Exception as e:
+        logger.error("Unable to connect to FTP server: try checking the host and port!")
+        raise e
+    try:
+        ftp.login(username, password)
+    except Exception as e:
+        logger.error("Unable to login to FTP server: try checking the username and password!")
+        raise e
+    return ftp
+
+
+# transport paramaters can include any extra parameters that you want to be passed into FTP_TLS
+def open(
+    path,
+    mode="r",
+    host=None,
+    user=None,
+    password=None,
+    port=DEFAULT_PORT,
+    transport_params=None,
+):
+    if not host:
+        raise ValueError("you must specify the host to connect to")
+    if not user:
+        user = getpass.getuser()
+    if not transport_params:
+        transport_params = {}
+    conn = _connect(host, user, port, password, transport_params)
+    mode_to_ftp_cmds = {
+        "r": ("RETR", "r"),
+        "rb": ("RETR", "rb"),
+        "w": ("STOR", "w"),
+        "wb": ("STOR", "wb"),
+        "a": ("APPE", "w"),
+        "ab": ("APPE", "wb")
+    }
+    try:
+        ftp_mode, file_obj_mode = mode_to_ftp_cmds[mode]
+    except KeyError:
+        raise ValueError(f"unsupported mode: {mode}")
+    ftp_mode, file_obj_mode = mode_to_ftp_cmds[mode]
+    fobj = conn.transfercmd(f"{ftp_mode} {path}").makefile(file_obj_mode)
```

Comment:
Two comments about this stemming from the [socket.makefile docs](https://docs.python.org/3/library/socket.html#socket.socket.makefile)

> the only supported mode values are 'r' (default), 'w' and 'b'

Do these other modes here actually work? It doesn't seem like they should

> Closing the file object returned by [makefile()]. 
 (https://docs.python.org/3/library/socket.html#socket.socket.makefile) won’t close the original socket unless all other file objects have been closed and [socket.close()](https://docs.python.org/3/library/socket.html#socket.close) has been called on the socket object.

I think we should make sure we are closing the underlying socket returned by `conn.transfercmd` here. On that note, do we also need to make sure we are closing the `conn` object?

Reply:
1. Interestingly, append works when we do the following: 
```
fobj = conn.transfercmd("APPE {file}").makefile("w")
fobj.write("something")
```
The result is that "something" is appended to the end of the file. [Here](https://www.serv-u.com/resources/tutorial/appe-stor-stou-retr-list-mlsd-mlst-ftp-command) are examples of ftp commands.

2. Good point. I will work on potentially returning a wrapper around the file object that closes both the underlying socket and the conn object. While in the docs [here](https://docs.python.org/3/library/ftplib.html#ftplib.FTP) it doesn't explicitly say that we should close to the conn object (which is of type ftplib.FTP), I think it is best to still close it, as we never reuse this connection.

Reply:
I have updated ftp.py to have the following:
```
socket = conn.transfercmd(f"{ftp_mode} {path}")
fobj = socket.makefile(file_obj_mode)
def full_close(self):
    self.orig_close()
    self.socket.close()
    self.conn.close()
fobj.orig_close = fobj.close
fobj.socket = socket
fobj.conn = conn
fobj.close = types.MethodType(full_close, fobj)
```

Now, the ftp connection, the socket, and the file will all be closed.